Support multi mount-string

bugwz opened the pull request:
<!-- 🎉 Thank you for contributing to minikube! 🎉 Here are some hints to get your PR merged faster:

1. Your PR title will be included in the release notes, choose it carefully
2. If the PR fixes an issue, add "fixes #<issue number>" to the description.
3. If the PR is a user interface change, please include a "before" and "after" example.
4. If the PR is a large design change, please include an enhancement proposal:
https://github.com/kubernetes/minikube/tree/master/enhancements
-->

When using `minikube start`, we can pass in multiple paths that we want to mount inside the container through the `mount-string` parameter, for example:

```bash
minikube start --mount --mount-string="/data01:/data01;/data02:/data02:Z,rshared"
```

bugwz committed:
Support multi mount-string

bugwz committed:
Fix multi mount error

medyagh commented:
/ok-to-test

medyagh commented:
@bugwz can you plz share Before/After this PR output of minikube?

bugwz commented:
@medyagh Below is the comparison information, please take a look～

**Before:**

```
[bugwz@iZj6cibr8kn3thfefwcyg3Z minikube]# ./out/minikube version
minikube version: v1.36.0
commit: 78989b99c29903d39ba5a8f2a0fef52d8b2c4497

[bugwz@iZj6cibr8kn3thfefwcyg3Z minikube]# ./out/minikube start --help
...
    --mount-string='/root:/minikube-host':
	The argument to pass the minikube mount command on start.
...

[bugwz@iZj6cibr8kn3thfefwcyg3Z minikube]$ ./out/minikube start --driver="podman" --mount --mount-string="/data01:/data01"
😄  minikube v1.36.0 on Centos 8.5.2111 (amd64)
✨  Using the docker driver based on user configuration

💣  Exiting due to PROVIDER_DOCKER_VERSION_EXIT_1: "docker version --format <no value>-<no value>:<no value>" exit status 1: Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
time="2025-07-12T12:19:12+08:00" level=error msg="XDG_RUNTIME_DIR directory \"/run/user/0\" is not owned by the current user"
📘  Documentation: https://minikube.sigs.k8s.io/docs/drivers/docker/

[bugwz@iZj6cibr8kn3thfefwcyg3Z minikube]$ ./out/minikube start --driver="podman" --mount --mount-string="/data01:/data01"
😄  minikube v1.36.0 on Centos 8.5.2111 (amd64)
❗  The minimum required version for podman is "4.9.0". your version is "3.3.1". minikube might not work. use at your own risk. To install latest version please see https://podman.io/getting-started/installation.html
✨  Using the podman driver based on user configuration

🧯  The requested memory allocation of 3072MiB does not leave room for system overhead (total system memory: 3625MiB). You may face stability issues.
💡  Suggestion: Start minikube with less memory allocated: 'minikube start --memory=3072mb'

📌  Using Podman driver with root privileges
👍  Starting "minikube" primary control-plane node in "minikube" cluster
🚜  Pulling base image v0.0.47 ...
💾  Downloading Kubernetes v1.33.2 preload ...
    > preloaded-images-k8s-v18-v1...:  348.11 MiB / 348.11 MiB  100.00% 10.32 M
    > gcr.io/k8s-minikube/kicbase...:  502.26 MiB / 502.26 MiB  100.00% 6.70 Mi
E0712 12:20:35.359832   15917 cache.go:225] Error downloading kic artifacts:  not yet implemented, see issue #8426
🔥  Creating podman container (CPUs=2, Memory=3072MB) ...
🐳  Preparing Kubernetes v1.33.2 on Docker 28.1.1 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔗  Configuring bridge CNI (Container Networking Interface) ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

[bugwz@iZj6cibr8kn3thfefwcyg3Z minikube]$ sudo podman inspect minikube
[
    {
        "Id": "db2b5308a1f00f4fa9bd5192286f320a0fe4ae8ca636a1a29d386505599a5d7c",
        "Created": "2025-07-12T12:21:45.17550111+08:00",
        "Path": "/usr/local/bin/entrypoint",
        "Args": [
            "/sbin/init"
...
        "Mounts": [
            {
                "Type": "volume",
                "Name": "minikube",
                "Source": "/var/lib/containers/storage/volumes/minikube/_data",
                "Destination": "/var",
                "Driver": "local",
                "Mode": "",
                "Options": [
                    "exec",
                    "nosuid",
                    "nodev",
                    "rbind"
                ],
                "RW": true,
                "Propagation": "rprivate"
            },
            {
                "Type": "bind",
                "Source": "/lib/modules",
                "Destination": "/lib/modules",
                "Driver": "",
                "Mode": "",
                "Options": [
                    "rbind"
                ],
                "RW": false,
                "Propagation": "rprivate"
            },
            {
                "Type": "bind",
                "Source": "/data01",
                "Destination": "/data01",
                "Driver": "",
                "Mode": "",
                "Options": [
                    "rbind"
                ],
                "RW": true,
                "Propagation": "rprivate"
            }
        ],
...
```


**After:**
```bash
[bugwz@iZj6cibr8kn3thfefwcyg3Z minikube]$ ./out/minikube version
minikube version: v1.36.0
commit: 14c679030368174bde9cbd24753f8c141c6e6da0

[bugwz@iZj6cibr8kn3thfefwcyg3Z minikube]# ./out/minikube start --help
...
    --mount-string='/home/bugwz:/minikube-host':
	The argument to pass the minikube mount command on start, in a semicolon-separated format.
...

[bugwz@iZj6cibr8kn3thfefwcyg3Z minikube]$ ./out/minikube start --driver="podman" --mount --mount-string="/data01:/data01;/data02:/data02:Z,rshared"
😄  minikube v1.36.0 on Centos 8.5.2111 (amd64)
❗  The minimum required version for podman is "4.9.0". your version is "3.3.1". minikube might not work. use at your own risk. To install latest version please see https://podman.io/getting-started/installation.html
✨  Using the podman driver based on user configuration

🧯  The requested memory allocation of 3072MiB does not leave room for system overhead (total system memory: 3625MiB). You may face stability issues.
💡  Suggestion: Start minikube with less memory allocated: 'minikube start --memory=3072mb'

📌  Using Podman driver with root privileges
👍  Starting "minikube" primary control-plane node in "minikube" cluster
🚜  Pulling base image v0.0.47 ...
E0712 12:30:33.731925   33550 cache.go:225] Error downloading kic artifacts:  not yet implemented, see issue #8426
🔥  Creating podman container (CPUs=2, Memory=3072MB) ...
🐳  Preparing Kubernetes v1.33.2 on Docker 28.1.1 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔗  Configuring bridge CNI (Container Networking Interface) ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

[bugwz@iZj6cibr8kn3thfefwcyg3Z minikube]$ sudo podman inspect minikube
[root@iZj6cibr8kn3thfefwcyg3Z ~]# podman inspect minikube
[
    {
        "Id": "7cea00a807474ac6a111bc91ab33c9a857ca17831c1ba4df0521221bdca384b9",
        "Created": "2025-07-12T12:30:43.641329142+08:00",
        "Path": "/usr/local/bin/entrypoint",
        "Args": [
            "/sbin/init"
        ],
...
        "Mounts": [
            {
                "Type": "volume",
                "Name": "minikube",
                "Source": "/var/lib/containers/storage/volumes/minikube/_data",
                "Destination": "/var",
                "Driver": "local",
                "Mode": "",
                "Options": [
                    "exec",
                    "nosuid",
                    "nodev",
                    "rbind"
                ],
                "RW": true,
                "Propagation": "rprivate"
            },
            {
                "Type": "bind",
                "Source": "/lib/modules",
                "Destination": "/lib/modules",
                "Driver": "",
                "Mode": "",
                "Options": [
                    "rbind"
                ],
                "RW": false,
                "Propagation": "rprivate"
            },
            {
                "Type": "bind",
                "Source": "/data01",
                "Destination": "/data01",
                "Driver": "",
                "Mode": "",
                "Options": [
                    "rbind"
                ],
                "RW": true,
                "Propagation": "rprivate"
            },
            {
                "Type": "bind",
                "Source": "/data02",
                "Destination": "/data02",
                "Driver": "",
                "Mode": "",
                "Options": [
                    "rbind"
                ],
                "RW": true,
                "Propagation": "rshared"
            }
        ],
...
```

medyagh commented:
thank you @bugwz can you plz fix the lint issue ?
u can run "make lint" or look at output of https://github.com/kubernetes/minikube/actions/runs/16223086483/job/45836007332?pr=21056

medyagh commented on pkg/drivers/kic/oci/types_test.go:
```diff
@@ -96,14 +96,14 @@ func TestParseMountString(t *testing.T) {
-			Name:        "relative container path",
-			MountString: "/foo/bar:baz/bat:private",
-			ExpectErr:   true,
+			Name:         "relative container path",
```
please add more unit test cases and include some test cases with semi colomon

medyagh commented on cmd/minikube/cmd/start_flags.go:
```diff
@@ -173,7 +173,7 @@ func initMinikubeFlags() {
 	startCmd.Flags().StringP(containerRuntime, "c", constants.DefaultContainerRuntime, fmt.Sprintf("The container runtime to be used. Valid options: %s (default: auto)", strings.Join(cruntime.ValidRuntimes(), ", ")))
 	startCmd.Flags().Bool(createMount, false, "This will start the mount daemon and automatically mount files into minikube.")
-	startCmd.Flags().String(mountString, constants.DefaultMountDir+":/minikube-host", "The argument to pass the minikube mount command on start.")
+	startCmd.Flags().String(mountString, constants.DefaultMountDir+":/minikube-host", "The argument to pass the minikube mount command on start, in a semicolon-separated format.")
```
can add an example here for clarity

medyagh reviewed (changes requested):
please see comments and also fix lint issues (make lint)

bugwz committed:
fix errors and add test

bugwz commented:
@medyagh Thank you for the suggestion, it has been updated.

medyagh reviewed (commented):
@bugwz does this mean multi-mount is only available for when users "starts" minikube?

how about if the user uses "minikube mount" command ?

I feel like this is a super useful feature and we should add some documents about it in our website maybe in tutorials




https://minikube.sigs.k8s.io/docs/handbook/mount/
https://minikube.sigs.k8s.io/docs/commands/mount/

medyagh reviewed (changes requested):
I tried manually, the "mount command" also needs to be updated so it can support multiple mount folders, would you plz add that feature there as well ?

```
$ minikube mount ~/.asdf/:/tmp1/;~/.vscode/:/tmp2
📁  Mounting host path /Users/medya/.asdf/ into VM as /tmp1/ ...
    ▪ Mount type:   9p
    ▪ User ID:      docker
    ▪ Group ID:     docker
    ▪ Version:      9p2000.L
    ▪ Message Size: 262144
    ▪ Options:      map[]
    ▪ Bind Address: 192.168.106.1:59806
🚀  Userspace file server: ufs starting
✅  Successfully mounted /Users/medya/.asdf/ to /tmp1/


```

Manually checked only /tmp1 was inside minikube

bugwz commented:
@medyagh Thank you for your suggestion. I will continue to improve as per your advice. Stay tuned!😜

bugwz committed:
Support multi mount for mount

bugwz commented:
@medyagh 

Hello, I have submitted new changes, please conduct a code review (CR).

**The relevant changes are as follows:**
1. Added support for multiple mounts in the mount feature.
2. Modified documentation for mount and start.


**Design approach for minikube mount:**
1. Added Root and RootIdx variables to the Ufs struct of the 9p server to record the hostPaths list and index.
2. Added a root variable to the ufsFid struct of the 9p server to record the mapped hostPath.
3. Passed the hostPaths list as an argument when starting the 9p server.
4. When the mount -t 9p command is called sequentially, the 9p server processes the requests from the kernel by calling the 5. Attach function. The Attach function selects Root[RootIdx] based on the execution order and assigns it to the fid.root variable.
6. Subsequent file and directory operations will parse the value of the fid.root variable.

medyagh commented:
@nirs please review this

bugwz commented:
Is anyone following up?

nirs commented:
> When using `minikube start`, we can pass in multiple paths that we want to mount inside the container through the `mount-string` parameter, for example:
> 
> ```shell
> minikube start --mount --mount-string="/data01:/data01;/data02:/data02:Z,rshared"
> ```

--mount is not use now. For --mount-string we should not use non standard format like this.

A better way would be to use multiple --mount-string flags using cmd.Flags().StringArray():

    --mount-string /dir1:/mnt/dir1[:options]
    --mount-string /dir2:/mnt/dir2[:options]

This will be parsed as:

    []string{"/dir1:/mnt/dir1:options", "/dir2:/mnt/dir2:options"}

The advantage is not having to use non-standard separator for users, and having to write special code for parsing the flags.

We ware adding virtiofs support (currently in vfkit and krunkit), so we need to pass options that can work with virtiofs. This should work also with 9p (legacy mount for other drivers), and for podamn/docker we need the special Z option.

We have issues with how --mount-options and other flags like --mount-uid and --mount=guid will be handled. We want to depracate these flags in 1.37 and remove them in 1.38, so options will be specified per mount.

nirs commented:
> minikube mount ~/.asdf/:/tmp1/;~/.vscode/:/tmp2

We should not support such thing, we can accept multiple arguments:

    minikube mount "/a:/b" "/c:/d"

But do can we mount multiple mounts with single mount command? It makes sense if we can run single mount server serving multiple commands. If we actually start multiple servers it is better to support one mount.
